src/ucp/stream/stream_recv.c

```diff
@@ -30,110 +62,311 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_stream_recv_data_nb,
+    } else {
+        ucs_mpool_put_inline(rdesc);
+    }
+}
+
 UCS_PROFILE_FUNC_VOID(ucp_stream_data_release, (ep, data),
                       ucp_ep_h ep, void *data)
 {
     ucp_recv_desc_t *rdesc = ucp_stream_rdesc_from_data(data);
 
     UCP_THREAD_CS_ENTER_CONDITIONAL(&ep->worker->mt_lock);
 
-    if (ucs_unlikely(rdesc->flags & UCP_RECV_DESC_FLAG_UCT_DESC)) {
-        uct_iface_release_desc((char*)rdesc - sizeof(ucp_eager_sync_hdr_t));
-    } else {
-        ucs_mpool_put_inline(rdesc);
-    }
+    ucp_stream_rdesc_release(rdesc);
 
     UCP_THREAD_CS_EXIT_CONDITIONAL(&ep->worker->mt_lock);
 }
 
+static UCS_F_ALWAYS_INLINE ssize_t
+ucp_stream_rdata_unpack(void *rdata, size_t length, ucp_request_t *dst_req)
+{
+    /* Truncated error is not actual for stream, need to adjust */
+    size_t       valid_len = ucs_min((dst_req->recv.length -
+                                      dst_req->recv.state.offset),
+                                     length);
+    ucs_status_t status    = ucp_dt_unpack(dst_req->recv.datatype,
+                                           dst_req->recv.buffer,
+                                           dst_req->recv.length,
+                                           &dst_req->recv.state, rdata,
+                                           valid_len, UCP_RECV_DESC_FLAG_LAST);
+    if (ucs_likely(status == UCS_OK)) {
+        dst_req->recv.state.offset += valid_len;
+        ucs_trace_data("unpacked %zd stream data\n", valid_len);
+
+        return valid_len;
+    }
+
+    ucs_assert(status != UCS_ERR_MESSAGE_TRUNCATED);
+
+    return status;
+}
+
 static UCS_F_ALWAYS_INLINE ucs_status_t
-ucp_stream_am_handler(void *arg, void *data, size_t length, unsigned am_flags)
+ucp_stream_process_rdesc(ucp_recv_desc_t* rdesc, ucp_ep_ext_stream_t *ep_stream,
+                         ucp_request_t *req)
 {
-    const size_t            hdr_len         = sizeof(ucp_stream_am_hdr_t);
-    ucp_worker_h            worker          = arg;
-    ucp_ep_h                ep;
-    ucp_stream_am_hdr_t     *hdr;
-    ucp_recv_desc_t         *rdesc;
-    size_t                  recv_len;
-    khiter_t                hash_it;
-    ucs_status_t            status;
+    void    *rdata   = ucp_stream_rdesc_data(rdesc);
+    ssize_t unpacked = ucp_stream_rdata_unpack(rdata, rdesc->length, req);
+
+    if (ucs_unlikely(unpacked < 0)) {
+        return unpacked;
+    } else if (ucs_likely(unpacked == rdesc->length)) {
+        if (rdesc == ep_stream->rdesc) {
+            ep_stream->rdesc = NULL;
+        }
+        ucp_stream_rdesc_release(rdesc);
+    } else {
+        ucs_assert(unpacked < rdesc->length);
 
-    UCP_THREAD_CS_ENTER_CONDITIONAL(&worker->mt_lock);
+        /* Save partially handled rdesc */
+        rdesc->length   -= unpacked;
+        ep_stream->rdesc = rdesc;
+        /* TODO: can we avoid extra copying? */
+        memmove(rdata, UCS_PTR_BYTE_OFFSET(rdata, unpacked), rdesc->length);
```

Comment:
this is a no-go (performance).
pls save rdesc_offset in ucp_ep_ext_stream_t

Reply:
Then, in common case, we need to save a lot of offsets for all pointers were returned to user and do lookup on release data, let's keep simple implementation for now and do optimizations when all tests will be in the repository?

Reply:
just need to save another variable - "offset" - in the ep, which is the next offset to copy from the 1st desc in the queue

Reply:
Again, it does not solve the problem which I decribed, ucp can return a lot of data pointers to partially handled rdescs by ucp_stream_recv_data_nb and we need to handle ucp_stream_release_data correctly in any number and order.